Optimize cancel_allocation SC

peterlimg opened the pull request:
## Fixes
- closes #572
## Changes
- Optimize cancel_allocation SC from around `14ms` to about `3.8ms`. Alloc blobbers' stake pools save process is still slow, so the more blobbers an alloc has, the slower the cancel_allocation SC would be. The benchmark test seems only has 4 blobbers each alloc. So it could be slower if an allocation has 12 blobbers. One way to improve it is do not save rewards to all blobbers' stake pools at once. Instead, we can save the rewards into an `alloc_blobber_rewards` partition, and distribute the rewards to one blobber each round in `payFees` txn.  In this way, no matter how many blobbers each alloc has, the cancel_allocation SC execution time would be the same all the time. 
- Removed unnecessary blobber storage node loading and saving
- Fixed a mistake on saving stake pools. Rewards didn't saved correctly, there were coins distributing to pools after saving. 

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

peterlimg committed:
Fix cancel_allocation benchmark

peterlimg committed:
Remove alloc from blobber when populate challenge

peterlimg committed:
Remove unnecessary blobber saving

peterlimg committed:
Remove unnecessary update blobber event

peterlimg committed:
Fix missed stake pool reward saving

dabasov commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -1456,8 +1456,10 @@ func (sc *StorageSmartContract) cancelAllocationRequest(
+	ts := time.Now()
 	var alloc *StorageAllocation
 	alloc, err = sc.getAllocation(req.AllocationID, balances)
+	fmt.Println("get allocation", time.Since(ts))
```
we better clean up. fmt

dabasov reviewed (approved):
lgtm, clean up timing logic please

peterlimg committed:
Clean up debug logs

peterlimg committed:
Fix unit tests

Remove allocation and blobber allocation checking after finishing allocation, we optimized to do not do those immediately, instead, do it in payFee txns.

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | TRUE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | optmize/cancel-alloc  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |
